File: ios/RNSiriShortcuts.m
//
//  RNSiriShortcuts.m
//  RNSiriShortcuts
//
//  Created by Gustavo Parreira on 23/09/2018.
//

#import "RNSiriShortcuts.h"
#import "NSUserActivity+ShortcutOptions.h"
#import <React/RCTBridgeModule.h>
#import <React/RCTConvert.h>
#import <React/RCTEventEmitter.h>
#import <UIKit/UIKit.h>
#import <Intents/Intents.h>
#import <IntentsUI/IntentsUI.h>
#import <CoreSpotlight/CoreSpotlight.h>

enum MutationStatus {
    MutationStatusCancelled,
    MutationStatusAdded,
    MutationStatusUpdated,
    MutationStatusDeleted,
};

NSString *MutationStatusToString(enum MutationStatus status) {
    switch (status) {
        case MutationStatusCancelled:
            return @"cancelled";
        case MutationStatusAdded:
            return @"added";
        case MutationStatusUpdated:
            return @"updated";
        case MutationStatusDeleted:
            return @"deleted";
    }
}

@interface RNSiriShortcuts () <INUIAddVoiceShortcutViewControllerDelegate, INUIEditVoiceShortcutViewControllerDelegate>
@end

@implementation RNSiriShortcuts {
    RCTResponseSenderBlock _Nullable _presentShortcutCallback;
    NSObject * _Nullable _editingVoiceShortcut; // Keep support for iOS <9
    UIViewController * _Nullable _presenterViewController;
    BOOL _hasListeners;
}
@synthesize bridge;

RCT_EXPORT_MODULE();

- (INVoiceShortcut * _Nullable)editingVoiceShortcut
API_AVAILABLE(ios(12.0))
{
    if ([_editingVoiceShortcut isKindOfClass:[INVoiceShortcut class]]) {
        return (INVoiceShortcut *)_editingVoiceShortcut;
    }
    return nil;
}

- (void)setEditingVoiceShortcut:(INVoiceShortcut * _Nullable)shortcut
API_AVAILABLE(ios(12.0))
{
    if (shortcut != _editingVoiceShortcut) {
        _editingVoiceShortcut = shortcut;
    }
}

#pragma mark Class Properties

+ (BOOL)application:(UIApplication *)application
continueUserActivity:(NSUserActivity *)userActivity
 restorationHandler:
#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= 12000) /* __IPHONE_12_0 */
(nonnull void (^)(NSArray<id<UIUserActivityRestoring>> *_Nullable))restorationHandler {
#else
(nonnull void (^)(NSArray *_Nullable))restorationHandler {
#endif
    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
    [notificationCenter postNotificationName:@"shortcutReceived"
                                      object:nil
                                    userInfo:@{
        @"userInfo": RCTNullIfNil(userActivity.userInfo),
        @"activityType": userActivity.activityType,
    }];
    return YES;
}

#pragma mark RCTBridgeModule

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

#pragma mark Contructors

- (instancetype)init
{
    self = [super init];
    if (self) {
        NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
        [notificationCenter addObserver:self
                               selector:@selector(handleReceivedShortcutNotification:)
                                   name:@"shortcutReceived"
                                 object:nil];
    }
    return self;
}

#pragma mark Observers

- (void)handleReceivedShortcutNotification:(NSNotification *)notification
{
    NSDictionary *activityInfo = notification.userInfo;
    
    if (_hasListeners) {
        [self sendEventWithName:@"SiriShortcutListener"
                           body:activityInfo];
    }
}

#pragma mark RCTEventEmitter

- (void)startObserving
{
    _hasListeners = YES;
    
    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
    [notificationCenter addObserver:self
                           selector:@selector(handleReceivedShortcutNotification:)
                               name:@"shortcutReceived"
                             object:nil];
}

- (void)stopObserving
{
    _hasListeners = NO;
    
    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
    [notificationCenter removeObserver:self
                                  name:@"shortcutReceived"
                                object:nil];
}

- (NSArray<NSString *> *)supportedEvents
{
    return @[@"SiriShortcutListener"];
}

#pragma mark RCT_EXPORT_METHODS

RCT_EXPORT_METHOD(getInitialShortcut:(RCTPromiseResolveBlock)resolve
                  rejecter:(__unused RCTPromiseRejectBlock)reject)
{
    // Extract the NSUserActivity data from launchOptions
    NSDictionary *userActivityDictionary = bridge.launchOptions[UIApplicationLaunchOptionsUserActivityDictionaryKey];
    NSDictionary * _Nullable shortcutInfo = nil;
    
    // If there was no NSUserActivity in launchOptions, the app was not launched from a shortcut
    if (userActivityDictionary) {
        NSArray<NSString *> *activityTypes = [NSBundle mainBundle].infoDictionary[@"NSUserActivityTypes"];
        // Check if the NSUserActivity.activityType matches one of the NSUserActivityTypes defined in Info.plist
        // If not, it was not launched from a shortcut that we should handle
        if ([activityTypes containsObject:userActivityDictionary[UIApplicationLaunchOptionsUserActivityTypeKey]]) {
            NSUserActivity *userActivity = userActivityDictionary[@"UIApplicationLaunchOptionsUserActivityKey"];
            shortcutInfo = @{
                @"userInfo": RCTNullIfNil(userActivity.userInfo),
                @"activityType": userActivity.activityType,
            };
        }
    }
    
    resolve(RCTNullIfNil(shortcutInfo));
}

RCT_EXPORT_METHOD(clearAllShortcuts:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    if (@available(iOS 12.0, *)) {
        [NSUserActivity deleteAllSavedUserActivitiesWithCompletionHandler:^{
            resolve(nil);
        }];
    } else {
        reject(@"below_ios_12",
               @"Your device needs to be running iOS 12+ for this",
               nil);
    }
}

RCT_EXPORT_METHOD(clearShortcutsWithIdentifiers:(NSArray *)persistentIdentifiers
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    if (@available(iOS 12.0, *)) {
        [NSUserActivity deleteSavedUserActivitiesWithPersistentIdentifiers:persistentIdentifiers
                                                         completionHandler:^{
            resolve(nil);
        }];
    } else {
        reject(@"below_ios_12",
               @"Your device needs to be running iOS 12+ for this",
               nil);
    };
}

RCT_EXPORT_METHOD(donateShortcut:(NSDictionary *)options)
{
    NSUserActivity *activity = [[NSUserActivity alloc] initWithJSONOptions:options];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [UIApplication sharedApplication]
            .keyWindow
            .rootViewController
            .userActivity = activity;
    });
    [activity becomeCurrent];
}

RCT_EXPORT_METHOD(suggestShortcuts:(NSArray<NSDictionary *> *)shortcutOptionsArr)
{
    if (@available(iOS 12.0, *)) {
        NSMutableArray<INShortcut *> *suggestions = [NSMutableArray new];
        
        for (NSDictionary *options in shortcutOptionsArr) {
            NSUserActivity *activity = [[NSUserActivity alloc] initWithJSONOptions:options];
            [suggestions addObject:[[INShortcut alloc] initWithUserActivity:activity]];
        }
        
        [[INVoiceShortcutCenter sharedCenter] setShortcutSuggestions:suggestions];
    }
}

RCT_EXPORT_METHOD(presentShortcut:(NSDictionary *)options
                  callback:(RCTResponseSenderBlock)callback)
{
    if (@available(iOS 12.0, *)) {
        NSUserActivity *activity = [[NSUserActivity alloc] initWithJSONOptions:options];
    
        INShortcut *shortcut = [[INShortcut alloc] initWithUserActivity:activity];
        
        [self fetchRecordedVoiceShortcutsWithCompletion:^(NSArray<INVoiceShortcut *> * _Nullable voiceShortcuts,
                                                          NSError * _Nullable error) {
            if (voiceShortcuts == nil) {
                if (error != nil) {
                    NSString *errorMsg = [[NSString alloc] initWithFormat:@"failed to fetch recorded shortcuts: %@", error];
                    callback(@[
                        @{
                            @"status": MutationStatusToString(MutationStatusCancelled),
                            @"error": errorMsg,
                        }
                    ]);
                    return;
                }
                
                callback(@[
                    @{
                        @"status": MutationStatusToString(MutationStatusCancelled),
                        @"error": @"unknown error fetching recorded shortcuts",
                    }
                ]);
                return;
            }
            
            self->_presentShortcutCallback = callback;
            INVoiceShortcut *addedVoiceShortcut;
            for (INVoiceShortcut *voiceShortcut in voiceShortcuts) {
                NSUserActivity * _Nullable userActivity = voiceShortcut.shortcut.userActivity;
                if (userActivity != nil &&
                    [userActivity.activityType isEqualToString:activity.activityType]) {
                    addedVoiceShortcut = voiceShortcut;
                    break;
                }
            }
            
            dispatch_async(dispatch_get_main_queue(), ^{
                if (addedVoiceShortcut != nil) {
                    // The shortcut was already added, so we present a form to edit it
                    [self setEditingVoiceShortcut:addedVoiceShortcut];
                    INUIEditVoiceShortcutViewController *presenterViewController = [[INUIEditVoiceShortcutViewController alloc] initWithVoiceShortcut:addedVoiceShortcut];
                    [presenterViewController setDelegate:self];
                    self->_presenterViewController = presenterViewController;
                } else {
                    // The shortcut was not added yet, so present a form to add it
                    INUIAddVoiceShortcutViewController *presenterViewController = [[INUIAddVoiceShortcutViewController alloc] initWithShortcut:shortcut];
                    [presenterViewController setDelegate:self];
                    self->_presenterViewController = presenterViewController;
                }
                self->_presenterViewController.modalPresentationStyle = UIModalPresentationFormSheet;
                UIViewController *rootViewController = [UIApplication sharedApplication].keyWindow.rootViewController;
                [rootViewController presentViewController:self->_presenterViewController
                                                 animated:true
                                               completion:nil];
            });
        }];
    }
}

RCT_EXPORT_METHOD(getShortcuts:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    if (@available(iOS 12.0, *)) {
        [self fetchRecordedVoiceShortcutsWithCompletion:^(NSArray<INVoiceShortcut *> * _Nullable voiceShortcuts,
                                                          NSError * _Nullable error) {
            if (error != nil) {
                reject(@"get_shortcuts_failure",
                       error.localizedDescription,
                       error);
                return;
            }
            
            if (voiceShortcuts == nil) {
                reject(@"get_shortcuts_failure",
                       @"An unknown error has occurred. Please open an issue on GitHub.",
                       nil);
                return;
            }
            
            NSMutableArray<NSDictionary<NSString *, id> *> *result = [NSMutableArray new];
            for (INVoiceShortcut *voiceShortcut in voiceShortcuts) {
                NSDictionary<NSString *, id> * _Nullable options = nil;
                
                NSUserActivity * _Nullable userActivity = voiceShortcut.shortcut.userActivity;
                if (userActivity != nil) {
                    options = @{
                        @"activityType": userActivity.activityType,
                        @"title": RCTNullIfNil(userActivity.title),
                        @"requiredUserInfoKeys": RCTNullIfNil(userActivity.requiredUserInfoKeys),
                        @"userInfo": RCTNullIfNil(userActivity.userInfo),
                        @"needsSave": @(userActivity.needsSave),
                        @"keywords": RCTNullIfNil(userActivity.keywords),
                        @"persistentIndentifier": RCTNullIfNil(userActivity.persistentIdentifier),
                        @"isEligibleForHandoff": @(userActivity.isEligibleForHandoff),
                        @"isEligibleForSearch": @(userActivity.isEligibleForSearch),
                        @"isEligibleForPublicIndexing": @(userActivity.isEligibleForPublicIndexing),
                        @"expirationDate": RCTNullIfNil(userActivity.expirationDate),
                        @"webpageURL": RCTNullIfNil(userActivity.webpageURL),
                        @"isEligibleForPrediction": @(userActivity.isEligibleForPrediction),
                        @"suggestedInvocationPhrase": RCTNullIfNil(userActivity.suggestedInvocationPhrase),
                    };
                }
                
                [result addObject:@{
                    @"identifier": voiceShortcut.identifier.UUIDString,
                    @"phrase": voiceShortcut.invocationPhrase,
                    @"options": RCTNullIfNil(options),
                }];
            }
            resolve(result);
        }];
    }
}

#pragma mark Helpers

API_AVAILABLE(ios(12.0))
typedef void (^FetchRecordedVoiceShortcutsCompletionBlock)(NSArray<INVoiceShortcut *>  * _Nullable voiceShortcuts, NSError  * _Nullable error);
- (void)fetchRecordedVoiceShortcutsWithCompletion:(FetchRecordedVoiceShortcutsCompletionBlock)completion
API_AVAILABLE(ios(12.0))
{
    [[INVoiceShortcutCenter sharedCenter] getAllVoiceShortcutsWithCompletion:^(NSArray<INVoiceShortcut *> * _Nullable voiceShortcuts,
                                                                               NSError * _Nullable error) {
        if (error != nil) {
            completion(nil, error);
            return;
        }
        
        if (voiceShortcuts != nil) {
            completion(voiceShortcuts, nil);
            return;
        }
        
        completion(nil, nil);
    }];
}

- (void)dismissPresenterWithStatus:(enum MutationStatus)status
                     voiceShortcut:(nullable INVoiceShortcut *)shortcut
API_AVAILABLE(ios(12.0))
{
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self->_presenterViewController != nil) {
            [self->_presenterViewController dismissViewControllerAnimated:YES completion:nil];
            self->_presenterViewController = nil;
        }
        
        if (self->_presentShortcutCallback != nil) {
            NSString *invocationPhrase = nil;
            if (shortcut != nil) {
                invocationPhrase = shortcut.invocationPhrase;
            }
            self->_presentShortcutCallback(@[
                @{
                    @"status": MutationStatusToString(status),
                    @"phrase": RCTNullIfNil(invocationPhrase),
                }
            ]);
            self->_presentShortcutCallback = nil;
        }
    });
}

#pragma mark INUIAddVoiceShortcutViewControllerDelegate

- (void)addVoiceShortcutViewController:(nonnull INUIAddVoiceShortcutViewController *)controller
            didFinishWithVoiceShortcut:(nullable INVoiceShortcut *)voiceShortcut
                                 error:(nullable NSError *)error
API_AVAILABLE(ios(12.0))
{
    // Shortcut was added
    [self dismissPresenterWithStatus:MutationStatusAdded
                       voiceShortcut:voiceShortcut];
}

- (void)addVoiceShortcutViewControllerDidCancel:(nonnull INUIAddVoiceShortcutViewController *)controller
API_AVAILABLE(ios(12.0))
{
    // Adding shortcut cancelled
    [self dismissPresenterWithStatus:MutationStatusCancelled
                       voiceShortcut:nil];
}

#pragma mark INUIEditVoiceShortcutViewControllerDelegate

- (void)editVoiceShortcutViewController:(nonnull INUIEditVoiceShortcutViewController *)controller
   didDeleteVoiceShortcutWithIdentifier:(nonnull NSUUID *)deletedVoiceShortcutIdentifier
API_AVAILABLE(ios(12.0))
{
    // Shortcut was deleted
    INVoiceShortcut * _Nullable deletedShortcut = [self editingVoiceShortcut];
    [self dismissPresenterWithStatus:MutationStatusDeleted
                       voiceShortcut:deletedShortcut];
    [self setEditingVoiceShortcut:nil];
}

- (void)editVoiceShortcutViewController:(nonnull INUIEditVoiceShortcutViewController *)controller
                 didUpdateVoiceShortcut:(nullable INVoiceShortcut *)voiceShortcut
                                  error:(nullable NSError *)error
API_AVAILABLE(ios(12.0))
{
    // Shortcut was updated
    [self dismissPresenterWithStatus:MutationStatusUpdated voiceShortcut:voiceShortcut];
    [self setEditingVoiceShortcut:nil];
}

- (void)editVoiceShortcutViewControllerDidCancel:(nonnull INUIEditVoiceShortcutViewController *)controller
API_AVAILABLE(ios(12.0))
{
    // Shortcut edit was cancelled
    [self dismissPresenterWithStatus:MutationStatusCancelled
                       voiceShortcut:nil];
    [self setEditingVoiceShortcut:nil];
}

@end

File: ios/RNSiriShortcuts.swift
//
//  ShortcutsModule.swift
//  RNSiriShortcuts
//
//  Created by Gustavo Parreira on 23/09/2018.
//  Copyright © 2018 Facebook. All rights reserved.
//

import Foundation
import UIKit
import Intents
import IntentsUI
import CoreSpotlight

enum VoiceShortcutMutationStatus: String {
    case cancelled = "cancelled"
    case added = "added"
    case updated = "updated"
    case deleted = "deleted"
}

@objc(ShortcutsModule)
open class ShortcutsModule: RCTEventEmitter, INUIAddVoiceShortcutViewControllerDelegate, INUIEditVoiceShortcutViewControllerDelegate {
    var hasListeners: Bool = false
    
    var _editingVoiceShortcut: NSObject? // Keep support for iOS <9
    @available(iOS 12.0, *)
    var editingVoiceShortcut: INVoiceShortcut? {
        get {
            _editingVoiceShortcut as? INVoiceShortcut
        }
        set {
           _editingVoiceShortcut = newValue
        }
    }
    var presenterViewController: UIViewController?
    var presentShortcutCallback: RCTResponseSenderBlock?
    @objc public static var initialUserActivity: NSUserActivity?
    
    @objc public static func onShortcutReceived(userActivity: NSUserActivity) {
        let userInfo = userActivity.userInfo
        let activityType = userActivity.activityType
        
        NotificationCenter.default.post(name: NSNotification.Name(rawValue: "InitialUserActivity"), object: nil, userInfo: [
            "userInfo": userInfo as Any,
            "activityType": activityType
            ])
    }
    
    override init() {
        super.init()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(startedFromShortcut(notification:)),
            name: NSNotification.Name(rawValue: "InitialUserActivity"),
            object: nil
        )
    }
    
    override open func startObserving() {
        hasListeners = true
        
        if let userActivity = ShortcutsModule.initialUserActivity {
            ShortcutsModule.onShortcutReceived(userActivity: userActivity)
            ShortcutsModule.initialUserActivity = nil
        }
    }
    
    override open func stopObserving() {
        hasListeners = false
    }
    
    override static public func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override open func supportedEvents() -> [String]! {
        return ["SiriShortcutListener"]
    }
    
    @objc func startedFromShortcut(notification: NSNotification) {
        let userInfo = notification.userInfo?["userInfo"]
        let activityType = notification.userInfo?["activityType"]
        
        if (hasListeners) {
            sendEvent(withName: "SiriShortcutListener", body: [
                "userInfo": userInfo,
                "activityType": activityType
                ])
        }
    }
    
    @available(iOS 9.0, *)
    static func generateUserActivity(_ jsonOptions: Dictionary<String, Any>) -> NSUserActivity {
        let options = ShortcutOptions(jsonOptions)
        
        let activity = NSUserActivity(activityType: options.activityType)
        activity.title = options.title
        activity.requiredUserInfoKeys = options.requiredUserInfoKeys
        activity.userInfo = options.userInfo
        activity.needsSave = options.needsSave
        activity.keywords = Set(options.keywords ?? [])
        activity.isEligibleForHandoff = options.isEligibleForHandoff
        activity.isEligibleForSearch = options.isEligibleForSearch
        activity.isEligibleForPublicIndexing = options.isEligibleForPublicIndexing
        activity.expirationDate = options.expirationDate
        
        if let urlString = options.webpageURL {
            activity.webpageURL = URL(string: urlString)
        }
        
        if #available(iOS 12.0, *) {
            activity.isEligibleForPrediction = options.isEligibleForPrediction
            activity.suggestedInvocationPhrase = options.suggestedInvocationPhrase
            if let identifier = options.persistentIdentifier {
                activity.persistentIdentifier = NSUserActivityPersistentIdentifier(identifier)
            }
        }
        
        let attributes = CSSearchableItemAttributeSet(itemContentType: options.contentType as String)
        if let description = options.contentDescription {
            attributes.contentDescription = description
        }
        activity.contentAttributeSet = attributes
        
        return activity
    }
    
    @available(iOS 9.0, *)
    @objc func donateShortcut(_ jsonOptions: Dictionary<String, Any>) {
        let activity = ShortcutsModule.generateUserActivity(jsonOptions)
        DispatchQueue.main.async {
            UIApplication.shared.keyWindow!.rootViewController!.userActivity = activity
        }
        activity.becomeCurrent()
    }
    
    @available(iOS 12.0, *)
    @objc func suggestShortcuts(_ jsonArray: Array<Dictionary<String, Any>>) {
        let suggestions = jsonArray.map { (_ options) -> INShortcut in
            let activity = ShortcutsModule.generateUserActivity(options)
            return INShortcut(userActivity: activity)
        }
        
        // Suggest the shortcuts.
        INVoiceShortcutCenter.shared.setShortcutSuggestions(suggestions)
    }
    
    @objc func clearAllShortcuts(_ resolve: @escaping RCTPromiseResolveBlock,
                                 rejecter reject: RCTPromiseRejectBlock) -> Void {
        if #available(iOS 12.0, *) {
            NSUserActivity.deleteAllSavedUserActivities {
                resolve(nil)
            }
        } else {
            reject("below_ios_12", "Your device needs to be running iOS 12+ for this", nil)
        }
    }
    
    @objc func clearShortcutsWithIdentifiers(_ persistentIdentifiers: Array<String>,
                                             resolver resolve: @escaping RCTPromiseResolveBlock,
                                             rejecter reject: RCTPromiseRejectBlock) -> Void {
        if #available(iOS 12.0, *) {
            let persistentIdentifierArr = persistentIdentifiers.map {
                NSUserActivityPersistentIdentifier($0)
            }
            
            NSUserActivity.deleteSavedUserActivities(withPersistentIdentifiers: persistentIdentifierArr,
                                                     completionHandler: { resolve(nil) })
        } else {
            reject("below_ios_12", "Your device needs to be running iOS 12+ for this", nil)
        }
    }
    
    @available(iOS 12.0, *)
    @objc func getShortcuts(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
        self.fetchRecorderdVoiceShortcuts { voiceShortcuts, error in
            if let error = error {
                reject("get_shortcuts_failure", error.localizedDescription, error)
                return
            }
            
            if let voiceShortcuts = voiceShortcuts {
                let result = voiceShortcuts.map { voiceShortcut -> [String: Any?] in
                    var options: [String: Any?]? = nil
                    if let userActivity = voiceShortcut.shortcut.userActivity {
                        options = [
                            "activityType": userActivity.activityType,
                            "title": userActivity.title,
                            "requiredUserInfoKeys": userActivity.requiredUserInfoKeys,
                            "userInfo": userActivity.userInfo,
                            "needsSave": userActivity.needsSave,
                            "keywords": userActivity.keywords,
                            "persistentIndentifier": userActivity.persistentIdentifier,
                            "isEligibleForHandoff": userActivity.isEligibleForHandoff,
                            "isEligibleForSearch": userActivity.isEligibleForSearch,
                            "isEligibleForPublicIndexing": userActivity.isEligibleForPublicIndexing,
                            "expirationDate": userActivity.expirationDate,
                            "webpageURL": userActivity.webpageURL,
                            "isEligibleForPrediction": userActivity.isEligibleForPrediction,
                            "suggestedInvocationPhrase": userActivity.suggestedInvocationPhrase
                        ]
                    }
                    
                    return [
                        "identifier": voiceShortcut.identifier.uuidString,
                        "phrase": voiceShortcut.invocationPhrase,
                        "options": options
                    ]
                }
                resolve(result)
                return
            }
            
            reject("get_shortcuts_failure", "An unknown error has occurred. Please open an issue on GitHub.", nil)
        }
    }
    
    @available(iOS 12.0, *)
    @objc func presentShortcut(_ jsonOptions: Dictionary<String, Any>, callback: @escaping RCTResponseSenderBlock) {
        let activity = ShortcutsModule.generateUserActivity(jsonOptions)

        let shortcut = INShortcut(userActivity: activity)
        
        self.fetchRecorderdVoiceShortcuts { voiceShortcuts, error in
            guard let voiceShortcuts = voiceShortcuts else {
                if let error = error {
                    callback([["status": VoiceShortcutMutationStatus.cancelled.rawValue,
                               "error": "failed to fetch recorded shortcuts: \(error)"]])
                    return
                }
                callback([["status": VoiceShortcutMutationStatus.cancelled.rawValue,
                           "error": "unknown error fetching recorded shortcuts"]])
                return
            }

            self.presentShortcutCallback = callback
            let addedVoiceShortcut = voiceShortcuts.first { (voiceShortcut) -> Bool in
              if let userActivity = voiceShortcut.shortcut.userActivity, userActivity.activityType == activity.activityType {
                return true
              }
              return false
            }

            DispatchQueue.main.async {
                if let addedVoiceShortcut = addedVoiceShortcut {
                    // The shortcut was already added, so we present a form to edit it
                    self.editingVoiceShortcut = addedVoiceShortcut
                    let presenterViewController = INUIEditVoiceShortcutViewController(voiceShortcut: addedVoiceShortcut)
                    presenterViewController.modalPresentationStyle = .formSheet
                    presenterViewController.delegate = self
                    self.presenterViewController = presenterViewController
                } else {
                    // The shortcut was not added yet, so present a form to add it
                    let presenterViewController = INUIAddVoiceShortcutViewController(shortcut: shortcut)
                    presenterViewController.modalPresentationStyle = .formSheet
                    presenterViewController.delegate = self
                    self.presenterViewController = presenterViewController
                }

                UIApplication.shared.keyWindow!.rootViewController!.present(self.presenterViewController!, animated: true, completion: nil)
            }
        }
    }

    @available(iOS 12.0, *)
    func dismissPresenter(_ status: VoiceShortcutMutationStatus, withShortcut voiceShortcut: INVoiceShortcut?) {
      DispatchQueue.main.async {
        self.presenterViewController?.dismiss(animated: true, completion: nil)
        self.presenterViewController = nil
        self.presentShortcutCallback?([
          ["status": status.rawValue, "phrase": voiceShortcut?.invocationPhrase]
        ])
        self.presentShortcutCallback = nil
      }
    }
    
    @available(iOS 12.0, *)
    public func addVoiceShortcutViewController(_ controller: INUIAddVoiceShortcutViewController,
                                               didFinishWith voiceShortcut: INVoiceShortcut?,
                                               error: Error?)
    {
        // Shortcut was added
        dismissPresenter(.added, withShortcut: voiceShortcut)
    }
    
    @available(iOS 12.0, *)
    public func addVoiceShortcutViewControllerDidCancel(_ controller: INUIAddVoiceShortcutViewController) {
        // Adding shortcut cancelled
        dismissPresenter(.cancelled, withShortcut: nil)
    }
    
    @available(iOS 12.0, *)
    public func editVoiceShortcutViewController(_ controller: INUIEditVoiceShortcutViewController,
                                                didUpdate voiceShortcut: INVoiceShortcut?,
                                                error: Error?)
    {
        // Shortcut was updated
        dismissPresenter(.updated, withShortcut: voiceShortcut)
        self.editingVoiceShortcut = nil
    }
    
    @available(iOS 12.0, *)
    public func editVoiceShortcutViewController(_ controller: INUIEditVoiceShortcutViewController,
                                                didDeleteVoiceShortcutWithIdentifier deletedVoiceShortcutIdentifier: UUID) {
        // Shortcut was deleted
        guard let editingVoiceShortcut = self.editingVoiceShortcut else {
            print("Could not find shortcut being edited.")
            dismissPresenter(.deleted, withShortcut: nil)
            return
        }
        dismissPresenter(.deleted, withShortcut: editingVoiceShortcut)
        self.editingVoiceShortcut = nil
    }
    
    @available(iOS 12.0, *)
    public func editVoiceShortcutViewControllerDidCancel(_ controller: INUIEditVoiceShortcutViewController) {
        // Shortcut edit was cancelled
        dismissPresenter(.cancelled, withShortcut: nil)
        self.editingVoiceShortcut = nil
    }

    @available(iOS 12.0, *)
    func fetchRecorderdVoiceShortcuts(completion: @escaping ([INVoiceShortcut]?, Error?) -> Void) {
        // Get all added voice shortcuts so we can make sure later if the presented shortcut is to be edited or added
        INVoiceShortcutCenter.shared.getAllVoiceShortcuts  { (voiceShortcutsFromCenter, error) in
            if let error = error {
                completion(nil, error)
                return
            }

            if let voiceShortcutsFromCenter = voiceShortcutsFromCenter {
                completion(voiceShortcutsFromCenter, nil)
                return
            }

            completion(nil, nil)
        }
    }
    
    // become current
    // resign current
    // invalidate
    
}
